sdk/python/v1beta1/kubeflow/katib/api/katib_client.py

```diff
@@ -411,9 +450,173 @@ def tune(
+                    setattr(lora_config, p_name, value)
+                    p_value.name = p_name
+                    experiment_params.append(p_value)
+                    trial_params.append(models.V1beta1TrialParameterSpec(name=p_name, reference=p_name))
+                elif p_value is not None:
+                    old_attr = getattr(lora_config, p_name, None)
+                    if old_attr is not None:
+                        value = type(old_attr)(p_value)
+                    setattr(lora_config, p_name, value)
+
+            # create init container spec.
+            init_container_spec = client.V1Container(
+                name=constants.STORAGE_INITIALIZER,
+                image=constants.STORAGE_INITIALIZER_IMAGE,
+                args=[
+                    "--model_provider",
+                    mp,
+                    "--model_provider_parameters",
+                    json.dumps(model_provider_parameters.__dict__, cls=utils.SetEncoder), 
+                    "--dataset_provider",
+                    dp,
+                    "--dataset_provider_parameters",
+                    json.dumps(dataset_provider_parameters.__dict__),
+                ],
+                volume_mounts=[constants.STORAGE_INITIALIZER_VOLUME_MOUNT],
+            )
+
+            lora_config = json.dumps(lora_config.__dict__, cls=utils.SetEncoder)
+            training_args = json.dumps(training_args.to_dict())
+            # create app container spec.
+            container_spec = client.V1Container(
+                name=constants.DEFAULT_PRIMARY_CONTAINER_NAME,
+                image=constants.TRAINER_TRANSFORMER_IMAGE,
+                args=[
+                    "--model_uri",
+                    model_provider_parameters.model_uri,
+                    "--transformer_type",
+                    model_provider_parameters.transformer_type.__name__,
+                    "--model_dir",
+                    constants.VOLUME_PATH_MODEL,
+                    "--dataset_dir",
+                    constants.VOLUME_PATH_DATASET,
+                    "--lora_config",
+                    f"'{lora_config}'",
+                    "--training_parameters",
+                    f"'{training_args}'",
+                ],
+                volume_mounts=[constants.STORAGE_INITIALIZER_VOLUME_MOUNT],
+                env=env if env else None,
+                env_from=env_from if env_from else None,
+                resources=resources_per_trial,
+            )
+
+            pod_spec = client.V1PodTemplateSpec(
+                metadata=models.V1ObjectMeta(
+                    annotations={"sidecar.istio.io/inject": "false"}
+                ),
+                spec=client.V1PodSpec(
+                    restart_policy="Never",
+                    containers=[container_spec],
+                    init_containers=[init_container_spec],
+                    volumes=[constants.STORAGE_INITIALIZER_VOLUME],
+                ),
+            )
```

Comment:
If we can, we should create helper function and re-use it in Training Operator + Katib.

Reply:
@andreyvelich I think we can reuse [get_container_spec](https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/training/utils/utils.py#L186) and [get_pod_template_spec](https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/training/utils/utils.py#L236) from the Training Operator. However, I have a small question. In Katib's PodSpec, we set the `restart_policy` to "Never," but in the Training Operator, we don't set it explicitly, so the default value is "Always." Can we maintain consistency between Katib and the Training Operator, or is there a specific reason for this difference?

Reply:
I think, `restart_policy=Never` is the only requirements for Job: https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-template, not for `PyTorchJob`

Reply:
@andreyvelich I have reused `get_container_spec` and `get_pod_template_spec` for the external models and datasets option. To enable this, a few parameters need to be added to these functions. Please review this PR:  https://github.com/kubeflow/training-operator/pull/2197. 
Additionally, I'm considering whether reusing them under custom objective function option to keep consistency. However, since users only need to install `kubeflow-training[huggingface]` when using external models and datasets, it might be better not to reuse them there. What are your thoughts?

Reply:
@andreyvelich Sorry I think I have misunderstood your meaning, so I initially tried to reuse functions from the Training Operator. Now, I've added some helper functions in `utils.py` for reuse, which are based on the helper functions in the Training Operator. Please review.